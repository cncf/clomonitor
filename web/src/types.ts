export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
