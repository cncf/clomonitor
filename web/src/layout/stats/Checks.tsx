import { REPORT_OPTIONS } from '../../data';
import { ReportOption, ReportOptionData } from '../../types';
import sortChecks from '../../utils/sortChecks';
import ProgressBarInLine from './ProgressBarInLine';

interface Props {
  title: string;
  data: { [key in ReportOption]?: number };
  onSelectCheck: (name: ReportOption) => void;
}

function getOptionInfo(key: ReportOption) {
  return REPORT_OPTIONS[key];
}

const Checks = (props: Props) => {
  const sortedChecks = sortChecks(props.data);

  return (
    <div className="col-12">
      <div className="fw-bold border-bottom pb-2 mb-3 lightText">{props.title}</div>
      <div className="mx-1 pt-1">
        {sortedChecks.map((check: ReportOption) => {
          const opt: ReportOptionData = getOptionInfo(check);
          return (
            <ProgressBarInLine
              key={`check_${opt.name}`}
              title={opt.shortName || opt.name}
              name={check}
              icon={opt.icon}
              value={props.data[check] as number}
              onSelectCheck={props.onSelectCheck}
            />
          );
        })}
      </div>
    </div>
  );
};

export default Checks;
