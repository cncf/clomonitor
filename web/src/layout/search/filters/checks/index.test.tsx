import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ReportOption } from '../../../../types';
import Checks from './index';

const mockOnChecksChange = jest.fn();
const mockOnChange = jest.fn();

const defaultProps = {
  activePassingChecks: [],
  activeNotPassingChecks: [],
  onChecksChange: mockOnChecksChange,
  onChange: mockOnChange,
};

describe('Checks', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<Checks {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders Checks', () => {
      render(<Checks {...defaultProps} />);

      expect(screen.getByText('Checks')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open checks modal' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open checks modal' })).toHaveTextContent('Add checks filters');
    });

    it('renders Checks with active options', () => {
      render(<Checks {...defaultProps} activePassingChecks={[ReportOption.Adopters]} />);

      expect(screen.getByText('Checks')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open checks modal' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open checks modal' })).toHaveTextContent('Edit checks filters');
    });

    it('renders Checks modal', async () => {
      render(<Checks {...defaultProps} />);

      const btn = screen.getByRole('button', { name: 'Open checks modal' });

      await userEvent.click(btn);

      expect(await screen.findByRole('dialog')).toBeInTheDocument();
      expect(screen.getByText('Checks filters')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Reset checks filters' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Reset checks filters' })).toBeDisabled();
      expect(screen.getByRole('button', { name: 'Apply filters' })).toBeInTheDocument();
    });

    it('calls onChecksChange to close modal filter', async () => {
      render(<Checks {...defaultProps} />);

      const btn = screen.getByRole('button', { name: 'Open checks modal' });

      await userEvent.click(btn);

      expect(await screen.findByRole('dialog')).toBeInTheDocument();

      const applyBtn = screen.getByRole('button', { name: 'Apply filters' });

      await userEvent.click(applyBtn);

      expect(mockOnChecksChange).toHaveBeenCalledTimes(1);
      expect(mockOnChecksChange).toHaveBeenCalledWith({ not_passing_check: [], passing_check: [] });

      expect(screen.queryByRole('dialog')).toBeNull();
    });

    it('calls onChange to click remove mini button', async () => {
      render(<Checks {...defaultProps} activePassingChecks={[ReportOption.Adopters]} />);

      const btn = screen.getByRole('button', { name: 'Open checks modal' });

      await userEvent.click(btn);

      expect(await screen.findByRole('dialog')).toBeInTheDocument();

      const removeBtn = screen.getByRole('button', { name: 'Remove passing Adopters check' });

      await userEvent.click(removeBtn);

      expect(mockOnChange).toHaveBeenCalledTimes(1);
      expect(mockOnChange).toHaveBeenCalledWith('passing_check', 'adopters', false);
    });
  });
});
