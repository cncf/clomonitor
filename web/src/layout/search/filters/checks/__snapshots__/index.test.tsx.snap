// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-row align-items-center text-primary categoryTitle"
  >
    <div
      class="fw-bold text-uppercase"
    >
      <small>
        Checks
      </small>
    </div>
  </div>
  <div>
    <button
      aria-label="Open checks modal"
      class="btn btn-outline-secondary btn-sm rounded-0 btn mt-3"
      type="button"
    >
      <span>
        Add checks filters
      </span>
    </button>
  </div>
</DocumentFragment>
`;
